### ISSUE TITLE
Inquiry about MedVAE training

### ISSUE THREAD

**Issue by Masaaki-75** [StanfordMIMI/MedVAE#1]
Great work and thanks for open-sourcing!

After reading your preprint I noticed that following details about MedVAE training seems missing. Could you please share some insights? 
- How are the images from different modalities preprocessed before being fed into the network? For example, the pixel values of a CT slice typically span several orders of magnitude (say [-1000, 3000]), while those of an MR image or FFDM are very different. How are they normalized?
- I am curious why the reconstruction loss (like L1 or L2, as a common practice in general VAE training) is not applied for MedVAE training. What benefits would we get when we drop that loss?

**Comment by ashwinkumargb** [comment]
Hi @Masaaki-75,

Thank you for taking the time to read our preprint! We are happy to clarify your comments.

- We normalize the CT scans from [-1000, 1000] to [0,1] using linear scaling ([code](https://github.com/StanfordMIMI/MedVAE/blob/34f7b58c1c9856293251346d5bd4598e93b719f2/medvae/utils/loaders.py#L106)). For MRI, we use min-max normalization to rescale intensities between [0,1] ([code](https://github.com/StanfordMIMI/MedVAE/blob/34f7b58c1c9856293251346d5bd4598e93b719f2/medvae/utils/loaders.py#L84)). After [0,1] normalization, we applied torchvision [normalize](https://github.com/StanfordMIMI/MedVAE/blob/34f7b58c1c9856293251346d5bd4598e93b719f2/medvae/utils/loaders.py#L84) to scale the values between [-1,1].  Although this was not included in the paper, we performed informal ablations testing different normalization strategies and training configurations (single vs. multi-modal training). Our current joint CT + MRI training approach for 3D MedVAE led to more enriched latent representations, which proved beneficial for downstream tasks.
- This is a great point! To clarify, in addition to the perceptual loss and patch-based adversarial objective, we **also** implemented an **L1 loss** (MedVAE [preprint](https://arxiv.org/abs/2502.14753); Page 13). This approach aligns with the methodology used in the Stable Diffusion [paper](https://arxiv.org/abs/2112.10752). As noted in their work, this combination helps ensure that "reconstructions [remain] confined to the image manifold, enforcing local realism and avoid[ing] blurriness introduced by relying solely on pixel-space losses such as L2 or L1 objectives." 

Please let us know if you have any further questions! We are also working on adding fine-tuning capabilities to the codebase, which will include additional training details.

**Comment by Masaaki-75** [comment]
Thank you for the clarification! Things now make sense to me. Love to see the fine-tuning code!

> * To clarify, in addition to the perceptual loss and patch-based adversarial objective, we **also** implemented an **L1 loss** (MedVAE [preprint](https://arxiv.org/abs/2502.14753); Page 13)

BTW, I guess you are referring to the L2 penalty imposed on the BiomedCLIP embeddings?

**Comment by ashwinkumargb** [comment]
Actually, sorry I am not referring to the L2 penalty, that is for the BiomedCLIP embeddings. We don't explicitly mention it in the paper, but here is some code, hopefully this helps!
```
# Perceptual loss
# Absolute Error
rec_loss = torch.abs(inputs.contiguous() - reconstructions.contiguous()) # L1 LOSS

# Perceptual Error (dim = [bsz x 1 x 1 x1])
p_loss = self.perceptual_loss(inputs.contiguous(), reconstructions.contiguous())
rec_loss = rec_loss + self.perceptual_weight * p_loss
nll_loss = (rec_loss / torch.exp(self.logvar) + self.logvar).sum() / bsz
```
